src/rust/engine/src/rule_graph.rs

```diff
@@ -258,183 +250,91 @@ impl<'t> GraphMaker<'t> {
-                "no rule was available to compute {} for {}",
-                type_constraint_str(input.clone()),
-                type_str(subject_type.clone())
-              ),
-            );
-            was_unfulfillable = true;
-          } else {
-            add_rules_to_graph(
-              &mut rules_to_traverse,
-              &mut rule_dependency_edges,
-              &mut unfulfillable_rules,
-              &entry,
-              SelectKey::JustSelect(Select::without_variant(*input)),
-              rules_or_literals_for_selector,
-            );
-          }
-        }
-      }
-      // TODO handle snapshot rules
-      if !was_unfulfillable {
-        // NB: In this case there were no selectors
-        add_rules_to_graph(
-          &mut rules_to_traverse,
-          &mut rule_dependency_edges,
-          &mut unfulfillable_rules,
-          &entry,
-          SelectKey::Nothing,
-          vec![],
-        );
-      }
-    }
-    RuleGraph {
-      root_subject_types: self.root_subject_types.clone(),
-      rule_dependency_edges: rule_dependency_edges,
-      unfulfillable_rules: unfulfillable_rules,
-      unreachable_rules: vec![],
-    }
-  }
-
-  fn _remove_unfulfillable_rules_and_dependents(&self, rule_graph: &mut RuleGraph) {
-    // Removes all unfulfillable rules transitively from the roots and the dependency edges.
-    //
-    // Takes the current root rule set and dependency table and removes all rules that are not
-    // transitively fulfillable.
+    rule_dependency_edges: &mut RuleDependencyEdges,
+    unfulfillable_rules: &mut UnfulfillableRuleMap,
+    entry: EntryWithDeps,
+  ) -> bool {
+    // If the entry has not been visited before, store a placeholder in the unfulfillable rules map
+    // and then visit its children. Otherwise, we're done.
     //
-    // Deforestation. Leaping from tree to tree.
-
-    let mut removal_traversal: VecDeque<_> =
-      rule_graph.unfulfillable_rules.keys().cloned().collect();
-    // could experiment with doing this for each rule added and deduping the traversal list
-    while let Some(unfulfillable_entry) = removal_traversal.pop_front() {
-      update_edges_based_on_unfulfillable_entry(
-        &mut rule_graph.rule_dependency_edges,
-        &mut rule_graph.unfulfillable_rules,
-        &mut removal_traversal,
-        &unfulfillable_entry,
-      );
-    }
+    // This prevents infinite recursion by shortcircuiting when an entry recursively depends on
+    // itself. It's totally fine for rules to be recursive: the recursive path just never
+    // contributes to whether the rule is satisfiable.
+    match (unfulfillable_rules.entry(entry.clone()), rule_dependency_edges.entry(entry.clone())) {
+      (hash_map::Entry::Vacant(_), hash_map::Entry::Vacant(re)) => {
+        // When a rule has not been visited before, we visit it by storing a placeholder in the
+        // rule dependencies map (to prevent infinite recursion).
+        re.insert(Default::default());
```

Comment:
I have a personal style preference to write this as `RuleEdges::default()` rather than `Default::default()` so it's easier to see at a glance what's happening :)